firestore/lib/src/data/restaurant_provider.dart

```diff
@@ -0,0 +1,168 @@
+    final restaurants = FirebaseFirestore.instance.collection('restaurants');
+    return restaurants.add({
+      'avgRating': restaurant.avgRating,
+      'category': restaurant.category,
+      'city': restaurant.city,
+      'name': restaurant.name,
+      'numRatings': restaurant.numRatings,
+      'photo': restaurant.photo,
+      'price': restaurant.price,
+    });
+  }
+
+  @override
+  void addRestaurantsBatch(List<Restaurant> restaurants) {
+    for (var restaurant in restaurants) addRestaurant(restaurant);
+  }
+
+  @override
+  Future<void> addReview(
+      {required String restaurantId, required Review review}) async {
+    final restaurant = await FirebaseFirestore.instance
+        .collection('restaurants')
+        .doc(restaurantId);
+    final newReview = await restaurant.collection('ratings').doc();
+
+    return FirebaseFirestore.instance
+        .runTransaction((Transaction transaction) async {
+      final DocumentSnapshot _restaurantSnapshot =
+          await transaction.get(restaurant);
+      final _restaurant = Restaurant.fromSnapshot(_restaurantSnapshot);
+      final newRatings = _restaurant.numRatings + 1;
+      final newAverage =
+          ((_restaurant.numRatings * _restaurant.avgRating) + review.rating) /
+              newRatings;
+
+      transaction.update(restaurant, {
+        'numRatings': newRatings,
+        'avgRating': newAverage,
+      });
+
+      transaction.set(newReview, {
+        'rating': review.rating,
+        'text': review.text,
+        'userName': review.userName,
+        'timestamp': review.timestamp,
+        'userId': review.userId,
+      });
+    });
+  }
+
+  @override
+  void loadAllRestaurants() {
+    final _querySnapshot = FirebaseFirestore.instance
+        .collection('restaurants')
+        .orderBy('avgRating', descending: true)
+        .limit(50)
+        .snapshots();
+
+    _querySnapshot.listen((event) {
+      final _restaurants = event.docs.map((DocumentSnapshot doc) {
+        return Restaurant.fromSnapshot(doc);
+      }).toList();
+
+      _allRestaurantsController.add(_restaurants);
+    });
+  }
+
+  @override
+  void loadFilteredRestaurants(Filter filter) {
+    Query collection = FirebaseFirestore.instance.collection('restaurants');
+    if (filter.category != null) {
+      collection = collection.where('category', isEqualTo: filter.category);
+    }
+    if (filter.city != null) {
+      collection = collection.where('city', isEqualTo: filter.city);
+    }
+    if (filter.price != null) {
+      collection = collection.where('price', isEqualTo: filter.price);
+    }
+    final _querySnapshot = collection
+        .orderBy(filter.sort ?? 'avgRating', descending: true)
+        .limit(50)
+        .snapshots();
+
+    _querySnapshot.listen((event) {
+      final _restaurants = event.docs.map((DocumentSnapshot doc) {
+        return Restaurant.fromSnapshot(doc);
+      }).toList();
+
+      _allRestaurantsController.add(_restaurants);
+    });
+  }
```

Comment:
I'm curious how this relates to `loadAllRestaurants`. If that is run first, this seems to be able to possibly save on downloads and use in-memory filtering?

Of course, all of this depends on our anticipated collection size.

Reply:
Alternatively, and I'm just realizing this now, this sample may not be setting out to always use Firestore the most strategically, and instead, demonstrate a menu of ways to load documents from a collection. If so, then I would guess most of my comments here are actually of fairly low value - but in that scenario, again, we should spell out that the sample is more about its individual methods and not so much about its higher level strategy.

Reply:
That second comment is quite a nice assumption, but the _truth_ is that I just copied most of the functions from the existing FriendlyEats demo. I haven't thought much about it either way. It is my literal job to think about these things and I appreciate the thorough review. I don't know what my next steps will be here, but there is good food for thought think about. I'll start with writing a README / docstrings, though.